implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java

```diff
@@ -52,185 +52,226 @@
+                    if (kid.equals(currentJwk.getKeyId())) {
+                        return PublicJsonWebKey.class.cast(currentJwk).getPublicKey();
+                    }
+                }
+            } else if (keys.size() == 1) {
+                return PublicJsonWebKey.class.cast(keys.get(0)).getPublicKey();
             }
-        } else {
-            kid = authContextInfo.getTokenKeyId();
+        } catch (Exception e) {
+            LOGGER.debug("Failed to read location as JWK(S)", e);
         }
+        return null;
+    }
 
-        PublicKey publicKey = null;
-
+    PublicKey tryAsLocalJwk(String content, String kid) {
         try {
-            log.debugf("Trying location as JWK(S)...");
+            JsonObject jwk = null;
 
-            if (httpsJwks != null) {
-                publicKey = getHttpJwk(httpsJwks, kid);
-            } else {
-                publicKey = getContentJwk(content, kid);
+            JsonObject jwks;
+            try (JsonReader reader = Json.createReader(new StringReader(content))) {
+                jwks = reader.readObject();
+            }
+            JsonArray keys = jwks.getJsonArray(JsonWebKeySet.JWK_SET_MEMBER_NAME);
+
+            if (keys != null) {
+                if (kid != null) {
+                    for (int i = 0; i < keys.size(); i++) {
+                        JsonObject currentJwk = keys.getJsonObject(i);
+                        if (kid.equals(currentJwk.getString(JsonWebKey.KEY_ID_PARAMETER, null))) {
+                            jwk = currentJwk;
+                            break;
+                        }
+                    }
+                } else if (keys.size() == 1) {
+                    jwk = keys.getJsonObject(0);
+                }
+            } else if (kid == null || kid.equals(jwks.getString(JsonWebKey.KEY_ID_PARAMETER, null))) {
+                jwk = jwks;
             }
-        } catch (Exception e) {
-            log.debug("Failed to read location as JWK(S)", e);
-        }
 
-        if (httpsJwks == null && publicKey != null && authContextInfo.getTokenKeyId() != null) {
-            // httpsJwks may refresh itself even if the jwksRefreshInterval is set to 0 as it checks HTTPS Cache-Control header
-            // so the public key will have to be created per request to ensure the rotation (if any) is effective.
-            verificationKey = publicKey;
+            if (jwk != null) {
+                return PublicJsonWebKey.Factory.newPublicJwk(jwk.toString()).getPublicKey();
+            }
+        } catch (Exception e) {
+            LOGGER.debug("Failed to read location as JWK(S)", e);
         }
 
-        return publicKey;
+        return null;
     }
 
-    PublicKey getHttpJwk(HttpsJwks httpsJwks, String kid) throws JoseException, IOException {
-        List<JsonWebKey> keys = httpsJwks.getJsonWebKeys();
+    private String getKid(JsonWebSignature jws) throws UnresolvableKeyException {
+        if (jws.getHeaders() == null)
```

Comment:
the JWS without the headers is invalid and Jose4J will fail earlier if it were the case, please remove this check